### ISSUE TITLE
0.31RC There was a problem with your question

### ISSUE THREAD

**Issue by rjurado01** [metabase/metabase#8894]
System info:

- Your browser and the version: Chrome 52.1
- Your operating system: Ubuntu 
- Your databases: MongoDB
- Metabase version: 0.31RC
- Metabase hosting environment: Docker
- Metabase internal database: Postgres

After update to 0.31RC I have this error when I make a question:

![image](https://user-images.githubusercontent.com/1703608/48188621-75490f80-e33e-11e8-949d-883a0ac08815.png)

I try to Resync-Table in admin panel but when I click on Settings button of table:

![image](https://user-images.githubusercontent.com/1703608/48188760-ce18a800-e33e-11e8-97d5-c485dfe86f94.png)

application loading appears and nothig happens (...admin/datamodel/database/2/table/9/settings):

![image](https://user-images.githubusercontent.com/1703608/48188814-e8eb1c80-e33e-11e8-8158-5b501e66eef8.png)

**Comment by senior** [comment]
It looks like `created_by_id` and `created_by_type` are missing. Were those removed, but due to you not being able to run sync, Metabase doesn't know about it yet?

When you visited the `admin/datamodel/database/2/table/9/settings` page and it started loading but never finished, were there any errors in the Metabase log? Were there any javascript errors in your browser's console?

**Comment by senior** [comment]
@camsaul It looks like the [column checking code](https://github.com/metabase/metabase/blob/master/src/metabase/driver/mongo/query_processor.clj#L607) is new for `0.31`. Since Mongo is schema-less, is it correct to have the expectation that all of the keys will be present in the results?

**Comment by camsaul** [comment]
@kdoh ?

**Comment by camsaul** [comment]
@senior we never fetch entire documents, we always tell Mongo what set of columns to return, so if we're returning columns that we didn't expect to get or vice versa there's a bug in the driver

It could be helpful to tweak that diff to use `clojure.data/diff` to show what's extra/missing. I think that's what we do in the `annotate` namespace.

**Comment by camsaul** [comment]
@rjurado01 can you use the browser console and include the full API response here? it would help a lot for debugging this

@senior I added some code to include original query, fully-preprocessed form, and native form in failed query executions as part of the refactor. It seems like our problem is that it's generating an incorrect native form given the fully-preprocessed MBQL form. Should be easy to write an exact test using those two forms

**Comment by senior** [comment]
@camsaul added the diffing like you suggested https://github.com/metabase/metabase/pull/8896

**Comment by rjurado01** [comment]
After update to `0.31` settings page works but error persist after Re-scan this table.

I have try it with other collection with less fields:

![image](https://user-images.githubusercontent.com/1703608/48252433-0aaed700-e405-11e8-845d-6e1eef262bdc.png)

The complete query is this:

```json
{
  "database_id": 2,
  "started_at": "2018-11-09T08:44:34.869Z",
  "json_query": {
    "database": 2,
    "type": "query",
    "query": {
      "source-table": 19
    },
    "parameters": [
      
    ],
    "constraints": {
      "max-results": 10000,
      "max-results-bare-rows": 2000
    }
  },
  "native": {
    "projections": [
      "_id",
      "name",
      "duration",
      "duration_ms",
      "exception",
      "file",
      "level",
      "level_index",
      "line",
      "message",
      "payload",
      "pid",
      "status",
      "tags",
      "thread",
      "time~~~default"
    ],
    "query": [
      {
        "$project": {
          "message": "$message",
          "time~~~default": {
            "$let": {
              "vars": {
                "column": "$time"
              },
              "in": "$$column"
            }
          },
          "level_index": "$level_index",
          "tags": "$tags",
          "exception": "$exception",
          "pid": "$pid",
          "name": "$name",
          "level": "$level",
          "status": "$status",
          "line": "$line",
          "duration_ms": "$duration_ms",
          "duration": "$duration",
          "_id": "$_id",
          "file": "$file",
          "thread": "$thread",
          "payload": "$payload"
        }
      },
      {
        "$project": {
          "message": "$message",
          "time~~~default": "$time~~~default",
          "level_index": "$level_index",
          "tags": "$tags",
          "exception": "$exception",
          "pid": "$pid",
          "name": "$name",
          "level": "$level",
          "status": "$status",
          "line": "$line",
          "duration_ms": "$duration_ms",
          "duration": "$duration",
          "_id": "$_id",
          "file": "$file",
          "thread": "$thread",
          "payload": "$payload"
        }
      },
      {
        "$limit": 2000
      }
    ],
    "collection": "semantic_logger",
    "mbql?": true
  },
  "status": "failed",
  "stacktrace": [
    "--> driver.mongo.query_processor$check_columns.invokeStatic(query_processor.clj:612)",
    "driver.mongo.query_processor$check_columns.invoke(query_processor.clj:607)",
    "driver.mongo.query_processor$execute_query.invokeStatic(query_processor.clj:643)",
    "driver.mongo.query_processor$execute_query.invoke(query_processor.clj:615)",
    "driver$fn__30407$G__30214__30414.invoke(driver.clj:106)",
    "query_processor$execute_query.invokeStatic(query_processor.clj:66)",
    "query_processor$execute_query.invoke(query_processor.clj:60)",
    "query_processor.middleware.mbql_to_native$mbql__GT_native$fn__32803.invoke(mbql_to_native.clj:36)",
    "query_processor.middleware.annotate$add_column_info$fn__31571.invoke(annotate.clj:298)",
    "query_processor.middleware.cumulative_aggregations$handle_cumulative_aggregations$fn__32271.invoke(cumulative_aggregations.clj:57)",
    "query_processor.middleware.limit$limit$fn__32756.invoke(limit.clj:17)",
    "query_processor.middleware.results_metadata$record_and_return_metadata_BANG_$fn__37439.invoke(results_metadata.clj:87)",
    "query_processor.middleware.format_rows$format_rows$fn__32746.invoke(format_rows.clj:26)",
    "query_processor.middleware.add_dimension_projections$add_remapping$fn__30964.invoke(add_dimension_projections.clj:236)",
    "query_processor.middleware.add_row_count_and_status$add_row_count_and_status$fn__31203.invoke(add_row_count_and_status.clj:15)",
    "driver.mongo$process_query_in_context$fn__61421$f__60472__auto____61423.invoke(mongo.clj:55)",
    "driver.mongo.util$_with_mongo_connection$fn__60465.invoke(util.clj:118)",
    "util.ssh$with_ssh_tunnel_STAR_.invokeStatic(ssh.clj:106)",
    "util.ssh$with_ssh_tunnel_STAR_.invoke(ssh.clj:91)",
    "driver.mongo.util$_with_mongo_connection.invokeStatic(util.clj:97)",
    "driver.mongo.util$_with_mongo_connection.invoke(util.clj:92)",
    "driver.mongo$process_query_in_context$fn__61421.invoke(mongo.clj:54)",
    "query_processor.middleware.driver_specific$process_query_in_context$fn__32507.invoke(driver_specific.clj:12)",
    "query_processor.middleware.resolve_driver$resolve_driver$fn__35249.invoke(resolve_driver.clj:15)",
    "query_processor.middleware.bind_effective_timezone$bind_effective_timezone$fn__31708$fn__31709.invoke(bind_effective_timezone.clj:9)",
    "util.date$call_with_effective_timezone.invokeStatic(date.clj:88)",
    "util.date$call_with_effective_timezone.invoke(date.clj:77)",
    "query_processor.middleware.bind_effective_timezone$bind_effective_timezone$fn__31708.invoke(bind_effective_timezone.clj:8)",
    "query_processor.middleware.store$initialize_store$fn__37454$fn__37455.invoke(store.clj:11)",
    "query_processor.store$do_with_new_store.invokeStatic(store.clj:34)",
    "query_processor.store$do_with_new_store.invoke(store.clj:30)",
    "query_processor.middleware.store$initialize_store$fn__37454.invoke(store.clj:10)",
    "query_processor.middleware.cache$maybe_return_cached_results$fn__32130.invoke(cache.clj:127)",
    "query_processor.middleware.catch_exceptions$catch_exceptions$fn__32178.invoke(catch_exceptions.clj:64)",
    "query_processor$process_query.invokeStatic(query_processor.clj:213)",
    "query_processor$process_query.invoke(query_processor.clj:209)",
    "query_processor$run_and_save_query_BANG_.invokeStatic(query_processor.clj:322)",
    "query_processor$run_and_save_query_BANG_.invoke(query_processor.clj:316)",
    "query_processor$fn__37627$process_query_and_save_execution_BANG___37632$fn__37633.invoke(query_processor.clj:354)",
    "query_processor$fn__37627$process_query_and_save_execution_BANG___37632.invoke(query_processor.clj:340)",
    "query_processor$fn__37651$process_query_and_save_with_max_BANG___37656$fn__37657.invoke(query_processor.clj:373)",
    "query_processor$fn__37651$process_query_and_save_with_max_BANG___37656.invoke(query_processor.clj:369)",
    "api.dataset$fn__41976$fn__41979.invoke(dataset.clj:45)",
    "api.common$fn__23138$invoke_thunk_with_keepalive__23143$fn__23144$fn__23145.invoke(common.clj:435)"
  ],
  "context": "ad-hoc",
  "error": "Error: mismatched columns in results! Expected: [:_id :name :duration :duration_ms :exception :file :level :level_index :line :message :payload :pid :status :tags :thread :time] Got: [:message :level_index :pid :name :level :_id :thread :time]",
  "row_count": 0,
  "running_time": 230,
  "preprocessed": {
    "database": 2,
    "type": "query",
    "query": {
      "source-table": 19,
      "fields": [
        [
          "field-id",
          290
        ],
        [
          "field-id",
          292
        ],
        [
          "field-id",
          298
        ],
        [
          "field-id",
          289
        ],
        [
          "field-id",
          337
        ],
        [
          "field-id",
          338
        ],
        [
          "field-id",
          299
        ],
        [
          "field-id",
          288
        ],
        [
          "field-id",
          339
        ],
        [
          "field-id",
          294
        ],
        [
          "field-id",
          293
        ],
        [
          "field-id",
          295
        ],
        [
          "field-id",
          300
        ],
        [
          "field-id",
          297
        ],
        [
          "field-id",
          291
        ],
        [
          "datetime-field",
          [
            "field-id",
            296
          ],
          "default"
        ]
      ],
      "limit": 2000
    },
    "constraints": {
      "max-results": 10000,
      "max-results-bare-rows": 2000
    },
    "driver": {
      
    },
    "settings": {
      
    }
  },
  "data": {
    "rows": [
      
    ],
    "cols": [
      
    ],
    "columns": [
      
    ]
  }
}
```

**Comment by camsaul** [comment]
Thanks @rjurado01 I'll take a look

**Comment by johnaldrintapia** [comment]
Hello guys, Just want to check if these issue already resolve. I also experiencing this error wherein I have a dynamic form builder which is possible that each record on our data in mongodb has different structure. Would that possible to put null/undefined on the missing field? So that we can preview the raw data without having this error "Query failure: Error: mismatched columns in results!"..

**Comment by rjurado01** [comment]
Posible related to #8929

**Comment by camsaul** [comment]
Fixed by #8952
